### PR TITLE
ci: Run a very slight continuous load against QA canary

### PR DESCRIPTION
**Pull request by def-**
We currently don't seem to have such a load, making it harder to detect certain kinds of problems before they hit customers.

Currently running in https://buildkite.com/materialize/nightlies/builds/6034#018d1be6-c555-49f2-a9cf-bd30cb122b47, will wait for it to finish before submitting

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by nrainer-materialize — APPROVED** [review]

File: test/canary-environment/models/table/schema.yml
Comment (philip-stoev) [inline test/canary-environment/models/table/schema.yml:1]: A more descriptive model name I think is in order here.

File: test/canary-load/mzcompose.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+from materialize.mzcompose.services.testdrive import Testdrive
+
+REGION = "aws/us-east-1"
+ENVIRONMENT = os.getenv("ENVIRONMENT", "production")
+USERNAME = os.getenv("PRODUCTION_SANDBOX_USERNAME", "infra+bot@materialize.com")
+APP_PASSWORD = os.environ["PRODUCTION_SANDBOX_APP_PASSWORD"]
+
+
+def hostname(c: Composition) -> str:
+    print("Obtaining hostname of staging instance ...")
```
Comment (philip-stoev) [inline test/canary-load/mzcompose.py:30]: We also do this bit in order mzcomposes , so maybe it is time to abstract it away as a utility function?

File: test/canary-load/mzcompose.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            no_reset=True,
+            materialize_url=f"postgres://{urllib.parse.quote(USERNAME)}:{urllib.parse.quote(APP_PASSWORD)}@{host}:6875/materialize",
+        )
+    ):
+        c.up("testdrive", persistent=True)
+
+        c.testdrive(
+            dedent(
+                """
+        > DELETE FROM qa_canary_environment.public_table.table
```
Comment (philip-stoev) [inline test/canary-load/mzcompose.py:64]: the """ opening and closing quotes are not aligned.

File: test/canary-environment/models/table/table.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+-- Use of this software is governed by the Business Source License
+-- included in the LICENSE file at the root of this repository.
+--
+-- As of the Change Date specified in that file, in accordance with
+-- the Business Source License, use of this software will be governed
+-- by the Apache License, Version 2.0.
+
+{{ config(materialized='table') }}
+
+CREATE TABLE {{ this }} (c int);
```
Comment (philip-stoev) [inline test/canary-environment/models/table/table.sql:12]: I think it is important to create a materialized view on top of this table , plus an index, just for the sake of putting load on the compute layer as well.

File: test/canary-load/mzcompose.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            current_time = time.time()
+
+            c.testdrive(
+                dedent(
+                    f"""
+            > SELECT 1
+            1
+
+            > INSERT INTO qa_canary_environment.public_table.table VALUES ({i})
+
```
Comment (philip-stoev) [inline test/canary-load/mzcompose.py:95]: Consider putting a comment here that all of the queries below are expected to return no results. Or convert them into `DISTINCT COUNT(*)` queries. The `DISTINCT` bit would cause a dataflow to be created to answer them

Or maybe the DISTINCT is an overkill, right now the dataflow is ensured by the presence of an `>` operator. - the EXPLAIN should show at least some queries with no `FastPath`

File: test/canary-load/mzcompose.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            user=USERNAME,
+            password=APP_PASSWORD,
+            port=6875,
+            ssl_context=ssl.create_default_context(),
+        )
+        cursor = conn.cursor()
+
+        cursor.execute("BEGIN")
+        cursor.execute(
+            "DECLARE subscribe CURSOR FOR SUBSCRIBE (SELECT * FROM qa_canary_environment.public_table.table)"
```
Comment (philip-stoev) [inline test/canary-load/mzcompose.py:80]: subscribe to the MV as well? In separate cursors?

File: test/feature-benchmark/scenarios.py
Comment (philip-stoev) [inline test/feature-benchmark/scenarios.py:1]: unrelated change?

File: ci/nightly/pipeline.template.yml [hunk 6]
```diff
@@ -1302,6 +1302,17 @@ steps:
         agents:
           queue: linux-x86_64-small
 
+  - id: qa-canary-load
```
Comment (philip-stoev) [inline ci/nightly/pipeline.template.yml:1305]: This is going to run for all the PRs as well , do we want that?

**Review by philip-stoev — APPROVED** [review]
Thank you, this is all much appreciated!

File: test/canary-load/mzcompose.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        while time.time() - start_time < args.runtime:
+            current_time = time.time()
+
+            c.testdrive(
+                dedent(
+                    f"""
+            > SELECT 1
+            1
+
+            > INSERT INTO qa_canary_environment.public_table.table VALUES ({i})
```
Comment (philip-stoev) [inline test/canary-load/mzcompose.py:94]: I suggest inserting 10s or 100s of rows instead here.

File: test/canary-load/mzcompose.py (hunk 1 above)
Comment (def-) [reply test/canary-load/mzcompose.py:30]: I actually wanted to do that and forgot, thanks for the reminder!

File: test/feature-benchmark/scenarios.py
Comment (def-) [reply test/feature-benchmark/scenarios.py:1]: Yes, wrong PR

File: ci/nightly/pipeline.template.yml (hunk 6 above)
Comment (def-) [reply ci/nightly/pipeline.template.yml:1305]: No, good catch.

File: test/feature-benchmark/scenarios.py
Comment (nrainer-materialize) [reply test/feature-benchmark/scenarios.py:1]: This might have broken the build: https://materializeinc.slack.com/archives/CTESPM7FU/p1705677816865339?thread_ts=1705597228.182579&cid=CTESPM7FU
